reco_utils/recommender/tfidf/tfidf_utils.py

```diff
@@ -0,0 +1,400 @@
+            df_clean (pd.DataFrame): Dataframe with cleaned text.
+        """
+        
+        # Initialize new dataframe to hold recommendation output
+        item_id = list()
+        title = list()
+
+        rec_rank = list()
+        rec_score = list()
+
+        rec_item_id = list()
+        rec_title = list()
+
+        for idx in range(0, len(self.recommendations)):
+            # Information about the item we are basing recommendations off of
+            rec_based_on = list(self.recommendations.keys())[idx]
+            tmp_item_id = str(df_clean.loc[df_clean[self.id_col] == rec_based_on][self.id_col].values[0])
+            tmp_title = str(df_clean.loc[df_clean[self.id_col] == rec_based_on][self.title_col].values[0])
+
+            # Iterate through top k recommendations
+            for i in range(0, self.k):
+                # Save to lists
+                item_id.append(tmp_item_id)
+                title.append(tmp_title)
+                rec_rank.append(str(i+1))
+                rec_score.append(self.recommendations[rec_based_on][i][0])
+                rec_item_id.append(self.recommendations[rec_based_on][i][1])
+                rec_title.append(self.recommendations[rec_based_on][i][2])
+
+        
+        # Save the output
+        output_dict = {self.id_col: item_id,
+                    'title': title,
+                    'rec_rank': rec_rank,
+                    'rec_score': rec_score,
+                    'rec_'+self.id_col: rec_item_id,
+                    'rec_title': rec_title}
+
+        # Convert to dataframe
+        self.top_k_recommendations = pd.DataFrame(output_dict)
+    
+    def recommend_top_k_items(self, df_clean):
+        """ Recommend k number of items similar to the item of interest.
+
+        Args:
+            df_clean (pd.DataFrame): Dataframe with cleaned text.
+        
+        Returns:
+            self.top_k_recommendations (pd.DataFrame): Dataframe containing id and title of top k recommendations for all items.
+        """
+        self.__create_full_recommendation_dictionary(df_clean)
+        self.__organize_results_as_tabular(df_clean)
+
+        return self.top_k_recommendations
+    
+    def __get_single_item_info(self, metadata, rec_id, verbose=True):
+        """ Get full information for a single recommended item.
+        
+        Args:
+            metadata (pd.DataFrame): Dataframe containing item info.
+            rec_id (str): Identifier for recommended item.
+            verbose (boolean): Set to True if you want to print basic info and URL.
+            
+        Results:
+            rec_info (pd.Series): Single row from dataframe containing recommended item info.
+        """
+        
+        # Return row
+        rec_info = metadata.iloc[int(np.where(metadata[self.id_col]==rec_id)[0])]
+
+        if verbose == True:
+            print('"'+ rec_info['title'] + '" (' + rec_info['publish_time'].rsplit('-')[0] + ')' + ' by ' + rec_info['authors'] + '.')
```

Comment:
there are a lot of specific assumptions about the columns in the dataframe. I would try to make this class as generalizable as possible and move specific covid dataset functionality to the notebook. 

It might help to make an assumption that there is only one column of text which this algorithm should work on, as well as an id column. The text can be cleaned, transformed to td-idf vectors, and then used for similarity based top-k estimation.

Reply:
`get_top_k_recommendations()` is now more generalized and not specific to the COVID-19 dataset.

Reply:
I have also generalized the class more by removing the need to specify a title column. The title column was more for convenience / sanity-checking purposes during development.